core/transaction-pool/graph/src/base_pool.rs

```diff
@@ -83,8 +83,7 @@ pub struct PruneStatus<Hash, Ex> {
 }
 
 /// Immutable transaction
-#[cfg_attr(test, derive(Clone))]
-#[derive(PartialEq, Eq)]
+#[derive(Clone, PartialEq, Eq)]
```

Comment:
Revalidate on light node means that we: (1) read all transactions from the pool (2) call `validate_transaction()` on remote node (3) resubmit updated transactions back to the pool. This at least requires `clone()` of `Transaction.data: Extrinsic` at (1), because it is an argument of `validate_transaction()`. And there have been clones of this data before - [here](https://github.com/paritytech/substrate/blob/c12e1d1f6ed7f59f58e30310078c6c6553ceb780/core/transaction-pool/graph/src/pool.rs#L311) for example.

What happens at (3) - basically I'm removing transactions that have been in the moment (1) from the pool. That's because their requires/provides could have been updated in revalidate call. Like they could became dependent on each other, or dependency could go away, or some transactions could became invalid. But what if between (1) and (3) new transaction has been inserted to the pool and this transaction depends on some tx from (1). I had couple of options - the most straightforward was to remove this transaction from the pool too. But we don't have revalidation results for this tx => I'm using requires/provides from previous `validate()` call (which is exactly what `base_pool::Transaction` stores). But on removal we only get the `Arc<Transaction>` (tbh - I do not know why - if we could get rid of `Arc`s that'd solve the problem) => that's why I need to clone contents of this `Arc`.

Reply:
We return `Arc<Transaction>` exactly to avoid cloning of the transaction content, and it has to exists in multiple places (in the pool and in the pending iterator for instance).

The code that clones `tx.data` could actually be optimized to first try to `Arc::try_unwrap` and only `clone` if there is another reference active (I suppose that 99% of cases would actually succeed in `try_unwrap`).

I'd really prefer to avoid making the transactions cloneable, they can be significant in size and it's going to be quite wasteful, not to mention a possible DoS vector. Do you think we could make a similar trick with `Arc::try_unwrap` in your new code and maybe avoid implementing `Clone` but rather adding some method that has more siginifcant name like `Transaction::duplicate()` to be able to document the rationale of not having `Clone`?